src/evict/evict_lru.c

```diff
@@ -1632,6 +1650,28 @@ __evict_walk_file(WT_SESSION_IMPL *session,
 	    QUEUE_FILLS_PER_PASS;
 
 	/*
+	 * If the tree is dead or we're near the end of the queue, fill the
+	 * remaining slots.
+	 */
+	if (F_ISSET(session->dhandle, WT_DHANDLE_DEAD))
+		target_pages = remaining_slots;
+
+	/*
+	 * Lookaside pages don't count toward the cache's dirty limit.
+	 *
+	 * Preferentially evict lookaside pages unless applications are stalled
+	 * on the dirty limit.  Once application threads are stalled by the
+	 * dirty limit, don't take any lookaside pages unless we're also up
+	 * against the total cache size limit.
+	 */
+	if (F_ISSET(btree, WT_BTREE_LOOKASIDE)) {
+		if (!F_ISSET(cache, WT_CACHE_EVICT_DIRTY_HARD))
+			target_pages = remaining_slots;
+		else if (!F_ISSET(cache, WT_CACHE_EVICT_CLEAN_HARD))
+			target_pages = 0;
+	}
```

Comment:
This is subtle and took a few attempts to get to this version.  It might need discussion and/or a longer comment.